src/sage/interfaces/singular.py: use GNU Info to read Singular's info

orlitzky opened the pull request:
Our Singular interface currently contains a hand-written parser for Singular's "info" file. This commit eliminates the custom parser in favor of launching GNU Info. GNU Info (or its superset, Texinfo) are widespread, portable, and easy to install on all of the systems we support, so in most cases this should be a "free" improvement.

The hand-written parser has several drawbacks:

* The extra code is a maintenance burden. We should not be wasting our time reimplementing standard tools.
* The custom parser is buggy. For example, it is supposed to raise a KeyError when documentation for a non-existent function is requested. However, the parser does not keep track of what section it's in, so, for example, get_docstring("Preface") returns the contents of the preface even though "Preface" is not a Singular function.
* The first time documentation is requested, the entire info file is loaded into a dictionary. This wastes a few megabytes of memory for the duration of the Sage session.
* The custom parser does not handle compression (GNU Info does transparently), and the end user or people packaging Singular may not be aware of that. If the system installation of Singular has a compressed info file, Sage won't be able to read it.

For contrast, the one downside to using GNU Info is that it adds a new runtime dependency to sagelib. To mitigate that, we do not technically require it, and instead raise a warning if the user (a) tries to read the Singular documentation and (b) has managed to find a system without GNU Info. Our singular_console() itself tries to launch GNU Info to display its interactive help, so the additional optional dependency is not so additional except in corner cases, such as a pypi installation of a subset of Sage linked against libsingular but without a full Singular install.

This was originally https://github.com/sagemath/sage/issues/32242

Since then,

* GNU Info has become a standard package in the sage distribution
* I'm now using `sage.features.info` to detect the "info" program, so nothing bad happens if it's not installed

Fixes https://github.com/sagemath/sage/issues/32242

orlitzky committed:
src/sage/interfaces/singular.py: use GNU Info to read Singular's info

Our Singular interface currently contains a hand-written parser for
Singular's "info" file. This commit eliminates the custom parser in
favor of launching GNU Info. GNU Info (or its superset, Texinfo) are
widespread, portable, and easy to install on all of the systems we
support, so in most cases this should be a "free" improvement.

The hand-written parser has several drawbacks:

* The extra code is a maintenance burden. We should not be wasting our
  time reimplementing standard tools.

* The custom parser is buggy. For example, it is supposed to raise a
  KeyError when documentation for a non-existent function is
  requested. However, the parser does not keep track of what section
  it's in, so, for example, get_docstring("Preface") returns the
  contents of the preface even though "Preface" is not a Singular
  function.

* The first time documentation is requested, the entire info file is
  loaded into a dictionary. This wastes a few megabytes of memory for
  the duration of the Sage session.

* The custom parser does not handle compression (GNU Info does
  transparently), and the end user or people packaging Singular may
  not be aware of that. If the system installation of Singular has a
  compressed info file, Sage won't be able to read it.

For contrast, the one downside to using GNU Info is that it adds a new
runtime dependency to sagelib. To mitigate that, we do not technically
require it, and instead raise a warning if the user (a) tries to read
the Singular documentation and (b) has managed to find a system
without GNU Info. Our singular_console() itself tries to launch GNU
Info to display its interactive help, so the additional optional
dependency is not so additional except in corner cases, such as a pypi
installation of a subset of Sage linked against libsingular but
without a full Singular install.

orlitzky committed:
src/sage/features/info.py: new feature for GNU Info

Before we can replace our hand-rolled parser for Singular's info file,
we need to be able to detect the "info" program from the GNU Info
package. The goal is to politely disable access to Singular's info if
GNU Info is unavailable. This avoids a hard dependency on GNU Info in
the sage library.

We require v7 or later because a bugfix in Texinfo v7.0.0 corrects the
exit code from "info" when the requested node is not found. This allows
us to separate the "expected failures" from the truly unexpected ones.

kwankyu commented on src/sage/interfaces/singular.py:
```diff
@@ -2408,15 +2350,65 @@ def get_docstring(name):
+    if not Info().is_present():
+        raise OSError("Error: GNU Info is not installed. Singular's "
+                      "documentation will not be available.")
+    import subprocess
```
just
```
OSError("GNU Info is not installed. Singular's "
```
?

orlitzky replied:
You're right, I copied that error message from some FFMpeg but there's no reason to say it again. Fixed.

orlitzky committed:
src/sage/libs/singular/function.pyx: use new get_docstring() features

The rewritten get_docstring() in sage.interfaces.singular supports
formatting the output as reStructuredText, so we no longer need to
do it "by hand" in the library interface.

orlitzky committed:
src/sage/misc/sageinspect.py: raise exceptions in _sage_getdoc_unformatted

Our _sage_getdoc_unformatted() catches all exceptions and turns them
into the empty string. The side effects of this are pretty bad: this
function is used to generate the help? strings in the sage REPL, and
if something goes wrong, the user will get a confusing blank docstring
rather than the error.

The change was introduced in issue 19671 as a workaround for a runtime
error raised inside flask. But we are no longer using flask, and
hopefully no other parts of the Sage library raise an exception during
the documentation build. We are about to find out!

orlitzky committed:
src/sage/algebras/letterplace: define singular functions locally

These modules use singular_function() from sage.libs.singular.function
to create top-level aliases for a few singular functions. The docbuild
however tries to build documentation for these functions, and it can
raise an error if GNU info is not installed. There's no good reason to
define them globally in the first place (they are only used locally),
so an easy fix is to move the imports and variables local.

kwankyu commented on src/sage/interfaces/singular.py:
```diff
@@ -2341,82 +2328,125 @@ def is_SingularElement(x):
+    - ``name`` -- a Singular function name
+    - ``prefix`` -- boolean (default: False); whether or not to include
+      the prefix stating that what follows is from the Singular
+      documentation.
```
```
``False``
```
here and below.

orlitzky replied:
fixed, thanks

orlitzky committed:
src/sage/interfaces/singular.py: False -> ``False`` in documentation

orlitzky committed:
src/sage/interfaces/singular.py: remove nonsense OUTPUT block

It looks like the OUTPUT block for singular_version() was copy/pasted
from somewhere else -- it doesn't make sense. We delete it; the
remaining short description of this function is sufficient.

kwankyu reviewed (approved):
LGTM now. Thanks.
